common/functions/src/scalars/dates/times.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::marker::PhantomData;
+
+use common_datavalues::chrono::Utc;
+use common_datavalues::prelude::*;
+use common_exception::Result;
+
+use crate::scalars::Function;
+
+#[derive(Clone, Debug)]
+pub struct TimeFunction<T> {
+    display_name: String,
+    t: PhantomData<T>,
+}
+
+pub trait NoArgTimeFunction {
+    fn execute() -> u32;
+}
+
+#[derive(Clone)]
+pub struct TimeSlot;
+
+impl NoArgTimeFunction for TimeSlot {
+    fn execute() -> u32 {
+        let utc = Utc::now();
+        (utc.timestamp_millis() / 1000 - 60 * 30) as u32
+    }
```

Comment:
That's wrong, what about the current time is exactly in half-hour mod?

Try using `utc.timestamp_millis() / 1000 / 1800 * 1800`

Reply:
> 1. Better to add docs for this function
> 2. How about to implement `timeSlots` in this pr ?

My fault.
I misunderstood the meaning of the timeSlot function